app/code/Magento/CatalogInventoryGraphQl/Model/Resolver/StockStatusProvider.php

```diff
@@ -45,9 +44,12 @@ public function resolve(Field $field, $context, ResolveInfo $info, array $value
         /* @var $product ProductInterface */
         $product = $value['model'];
 
-        $stockStatus = $this->stockStatusRepository->get($product->getId());
-        $productStockStatus = (int)$stockStatus->getStockStatus();
+        $stockStatus = $this->stockRegistry->getStockStatus(
+            $product->getId(),
+            $product->getStore()->getWebsiteId()
+        )->getStockStatus();
+        $isInStock = $stockStatus && $product->isSaleable();
```

Comment:
This bug is deeper that we are thinking. Looks like the root cause is somewhere in the `\Magento\CatalogInventory\Model\StockRegistryProvider::getStockStatus` method. Please debug to understand, why it does not take `saleable` into account. 
In other words `getStockStatus` should return correct stock status without calling `$product->isSaleable();`, because product should know nothing about its stock.